File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/engine/agent/TriageClassifier.kt
package org.wyrdsekai.app.engine.agent

import org.wyrdsekai.app.inference.ChatMessage
import org.wyrdsekai.app.inference.CompletionOptions
import org.wyrdsekai.app.inference.InferenceClient

/**
 * Classifies user input as SIMPLE or COMPLEX to route between fast local
 * inference (0.6B) and deeper household inference (7B+).
 *
 * Uses a three-tier approach:
 * 1. Heuristic fast-path (no inference): greetings, MUD commands, long input
 * 2. LLM classification (~2s on 0.6B): one-word response
 * 3. Fallback: SIMPLE (if classification fails)
 */
object TriageClassifier {

    enum class Tier { ROUTINE, SIMPLE, COMPLEX }

    // Acknowledgment patterns → ROUTINE (cheapest model)
    private val ACK_PATTERNS = setOf(
        "ok", "okay", "sure", "thanks", "thank you", "ty", "thx",
        "yes", "yeah", "yep", "yup", "ya", "no", "nope", "nah",
        "cool", "nice", "great", "awesome", "got it", "understood",
        "right", "correct", "fine", "alright", "k", "kk",
        "lol", "haha", "heh", "lmao", "rofl",
    )

    /**
     * Map tier to capability name for InferenceRouter.
     */
    fun tierToCapability(tier: Tier): String = when (tier) {
        Tier.ROUTINE -> "quick"
        Tier.SIMPLE -> "default"
        Tier.COMPLEX -> "reasoning"
    }

    // Greeting patterns (case-insensitive)
    private val GREETING_PATTERNS = setOf(
        "hi", "hello", "hey", "yo", "sup", "howdy", "hiya",
        "good morning", "good evening", "good afternoon",
        "what's up", "whats up", "wassup",
    )

    // MUD commands that don't need inference at all
    private val MUD_COMMANDS = setOf(
        "look", "l", "go", "move", "take", "get", "drop", "use",
        "inventory", "help", "socials",
        "nod", "smile", "laugh", "grin", "frown", "shrug", "sigh",
        "gasp", "blink", "wince", "wave", "bow", "clap", "dance",
        "stretch", "yawn", "pace", "fidget", "cry", "cheer", "groan",
        "blush", "ponder", "brood", "beam", "sulk", "hug", "thank",
        "agree", "disagree", "salute", "welcome",
    )

    private const val CLASSIFICATION_PROMPT = """Classify this message as SIMPLE or COMPLEX.
SIMPLE: greetings, yes/no, navigation, short answers, acknowledgments.
COMPLEX: questions needing thought, personal topics, creative requests, help requests, multi-step tasks.
Answer with one word only."""

    /**
     * Classify input complexity. Returns SIMPLE or COMPLEX.
     *
     * @param text The user's input text
     * @param inferenceClient Local inference client (0.6B)
     * @param inferenceBaseUrl Base URL for inference
     * @return Tier.SIMPLE or Tier.COMPLEX
     */
    suspend fun classify(
        text: String,
        inferenceClient: InferenceClient,
        inferenceBaseUrl: String,
    ): Tier {
        val trimmed = text.trim()
        val lower = trimmed.lowercase()

        // Heuristic fast-path: no inference needed
        val heuristic = heuristicClassify(lower, trimmed)
        if (heuristic != null) return heuristic

        // LLM classification
        return try {
            llmClassify(trimmed, inferenceClient, inferenceBaseUrl)
        } catch (_: Exception) {
            Tier.SIMPLE // Fallback: treat as simple if classification fails
        }
    }

    /**
     * Pure heuristic classification. Returns null if undecided (needs LLM).
     */
    internal fun heuristicClassify(lower: String, original: String): Tier? {
        // Strip trailing punctuation for matching
        val stripped = lower.replace(Regex("[.!?]+$"), "").trim()

        // Short greetings → ROUTINE
        if (stripped in GREETING_PATTERNS) return Tier.ROUTINE

        // Acknowledgments → ROUTINE
        if (stripped in ACK_PATTERNS) return Tier.ROUTINE

        // MUD commands (1-2 words only) → skip inference entirely
        val words = stripped.split("\\s+".toRegex())
        val wordCount = words.size
        if (wordCount <= 2 && words[0] in MUD_COMMANDS) return null

        // Very short (1-3 words, no question mark) → ROUTINE
        if (wordCount <= 3 && !original.contains("?")) return Tier.ROUTINE

        // Short (4-5 words, no question) → SIMPLE
        if (wordCount <= 5 && !original.contains("?")) return Tier.SIMPLE

        // Long input (> 30 words) → COMPLEX
        if (wordCount > 30) return Tier.COMPLEX

        // Contains question mark → likely needs thought
        if (original.contains("?") && wordCount > 8) return Tier.COMPLEX

        // Keywords indicating need for external knowledge or deep reasoning → COMPLEX
        val complexKeywords = listOf(
            "news", "today", "yesterday", "latest", "recent", "current",
            "happening", "happened", "explain", "analyze", "research",
            "summarize", "review", "compare", "investigate", "story about",
            "tell me about", "what do you think", "help me", "create a",
            "write a", "build a", "design a", "plan a",
        )
        if (complexKeywords.any { lower.contains(it) }) return Tier.COMPLEX

        // Undecided — needs LLM
        return null
    }

    /**
     * LLM-based classification. One inference call with tiny prompt.
     */
    private suspend fun llmClassify(
        text: String,
        inferenceClient: InferenceClient,
        inferenceBaseUrl: String,
    ): Tier {
        val messages = listOf(
            ChatMessage(role = "system", content = CLASSIFICATION_PROMPT),
            ChatMessage(role = "user", content = "Message: \"$text\""),
        )
        val response = inferenceClient.complete(
            baseUrl = inferenceBaseUrl,
            messages = messages,
            options = CompletionOptions(maxTokens = 8, temperature = 0.1),
        )
        val answer = response.content.trim().uppercase()
        return if (answer.contains("COMPLEX")) Tier.COMPLEX else Tier.SIMPLE
    }
}

File: clients/rn/src/engine/agent/TriageClassifier.ts
import {ChatMessage, CompletionOptions, ChatResponse} from '../../inference/types';

/**
 * Classifies user input as ROUTINE, SIMPLE, or COMPLEX to route between
 * model tiers: cheapest (greetings) → default → most capable (analysis).
 *
 * Uses a three-tier approach:
 * 1. Heuristic fast-path (no inference): greetings, MUD commands, long input
 * 2. LLM classification (~2s on 0.6B): one-word response
 * 3. Fallback: SIMPLE (if classification fails)
 */

export type Tier = 'routine' | 'simple' | 'complex';

export type InferFn = (
  messages: ChatMessage[],
  options?: CompletionOptions,
) => Promise<ChatResponse>;

// Greeting patterns (case-insensitive)
const GREETING_PATTERNS = new Set([
  'hi', 'hello', 'hey', 'yo', 'sup', 'howdy', 'hiya', 'heya',
  'good morning', 'good evening', 'good afternoon', 'good night',
  'morning', 'afternoon', 'evening', 'gm', 'gn',
  "what's up", 'whats up', 'wassup', 'wazzup',
]);

// Acknowledgment patterns → ROUTINE
const ACK_PATTERNS = new Set([
  'ok', 'okay', 'sure', 'thanks', 'thank you', 'ty', 'thx',
  'yes', 'yeah', 'yep', 'yup', 'ya', 'no', 'nope', 'nah',
  'cool', 'nice', 'great', 'awesome', 'got it', 'understood',
  'right', 'correct', 'fine', 'alright', 'k', 'kk',
  'lol', 'haha', 'heh', 'lmao', 'rofl',
]);

// MUD commands that don't need inference at all
const MUD_COMMANDS = new Set([
  'look', 'l', 'go', 'move', 'take', 'get', 'drop', 'use',
  'inventory', 'help', 'socials',
  'nod', 'smile', 'laugh', 'grin', 'frown', 'shrug', 'sigh',
  'gasp', 'blink', 'wince', 'wave', 'bow', 'clap', 'dance',
  'stretch', 'yawn', 'pace', 'fidget', 'cry', 'cheer', 'groan',
  'blush', 'ponder', 'brood', 'beam', 'sulk', 'hug', 'thank',
  'agree', 'disagree', 'salute', 'welcome',
]);

const CLASSIFICATION_PROMPT = `Classify this message as SIMPLE or COMPLEX.
SIMPLE: greetings, yes/no, navigation, short answers, acknowledgments.
COMPLEX: questions needing thought, personal topics, creative requests, help requests, multi-step tasks.
Answer with one word only.`;

/**
 * Map tier to capability name for InferenceRouter.
 */
export function tierToCapability(tier: Tier): string {
  switch (tier) {
    case 'routine': return 'quick';
    case 'simple': return 'default';
    case 'complex': return 'reasoning';
  }
}

/**
 * Pure heuristic classification. Returns null if undecided (needs LLM).
 */
export function heuristicClassify(text: string): Tier | null {
  const trimmed = text.trim();
  const lower = trimmed.toLowerCase();
  const stripped = lower.replace(/[.!?]+$/, '').trim();

  // Short greetings → ROUTINE
  if (GREETING_PATTERNS.has(stripped)) return 'routine';

  // Acknowledgments → ROUTINE
  if (ACK_PATTERNS.has(stripped)) return 'routine';

  // MUD commands (1-2 words only) → skip inference
  const words = stripped.split(/\s+/);
  const wordCount = words.length;
  if (wordCount <= 2 && MUD_COMMANDS.has(words[0] ?? '')) return null;

  // Very short (1-3 words, no question mark) → ROUTINE
  if (wordCount <= 3 && !trimmed.includes('?')) return 'routine';

  // Short (4-5 words, no question) → SIMPLE
  if (wordCount <= 5 && !trimmed.includes('?')) return 'simple';

  // Long input (> 30 words) → COMPLEX
  if (wordCount > 30) return 'complex';

  // Contains question mark → likely needs thought
  if (trimmed.includes('?') && wordCount > 8) return 'complex';

  // Keywords indicating need for external knowledge or deep reasoning → COMPLEX
  const complexKeywords = [
    'news', 'today', 'yesterday', 'latest', 'recent', 'current',
    'happening', 'happened', 'explain', 'analyze', 'research',
    'summarize', 'review', 'compare', 'investigate', 'story about',
    'tell me about', 'what do you think', 'help me', 'create a',
    'write a', 'build a', 'design a', 'plan a',
  ];
  if (complexKeywords.some(kw => lower.includes(kw))) return 'complex';

  // Undecided — needs LLM
  return null;
}

/**
 * Classify input complexity. Returns tier.
 */
export async function classify(text: string, infer: InferFn): Promise<Tier> {
  const trimmed = text.trim();

  // Heuristic fast-path: no inference needed
  const heuristic = heuristicClassify(trimmed);
  if (heuristic != null) return heuristic;

  // LLM classification
  try {
    return await llmClassify(trimmed, infer);
  } catch {
    return 'simple'; // Fallback
  }
}

/**
 * LLM-based classification. One inference call with tiny prompt.
 */
async function llmClassify(text: string, infer: InferFn): Promise<Tier> {
  const messages: ChatMessage[] = [
    {role: 'system', content: CLASSIFICATION_PROMPT},
    {role: 'user', content: `Message: "${text}"`},
  ];
  const response = await infer(messages, {maxTokens: 8, temperature: 0.1});
  const answer = response.content.trim().toUpperCase();
  return answer.includes('COMPLEX') ? 'complex' : 'simple';
}
